src/audio/src.c

```diff
@@ -777,42 +737,49 @@ static int src_copy(struct comp_dev *dev)
 	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
 			       source_list);
 
-	/* In some conversions the first copy run needs to pre-fill buffer
-	 * with sufficient amount of zeros if the min. output block length
-	 * is too short. It prevents xrun for the downstream component. In
-	 * successive copy executions the block length will jitter around the
-	 * nominal period length and xruns won't happen.
-	 */
-	if (cd->prefill && sink->free >= cd->prefill) {
-		tracev_src("src_copy(), need to "
-			   "pre-fill buffer, cd->prefill = %u", cd->prefill);
-		comp_update_buffer_produce(sink, cd->prefill);
-		cd->prefill = 0;
+	/* check for underrun */
+	if (source->avail == 0) {
+		trace_src_error("src_copy() error: Empty source buffer.");
+		comp_underrun(dev, source, 0, 0);
+		return -EIO;
 	}
 
-	/* Calculate needed amount of source buffer and sink buffer
-	 * for one SRC run. The blk_in and blk are minimum condition to
-	 * call copy. Copy can consume or produce a slightly larger block
-	 * with the rates where block sizes are not constant. E.g. for
-	 * 1 ms scheduling the blocks can be under or above 1 ms when the
-	 * SRC interval block size constraint prevents exact 1 ms blocks.
-	 */
-	need_source = cd->param.blk_in * dev->frame_bytes;
-	need_sink = cd->param.blk_out * dev->frame_bytes;
+	/* check for overrun */
+	if (sink->free == 0) {
+		trace_src_error("src_copy() error: Full sink buffer.");
+		comp_overrun(dev, sink, 0, 0);
+		return -EIO;
+	}
+
+	/* Get SRC parameters */
+	sp = &cd->param;
+	s1 = cd->src.stage1;
+	s2 = cd->src.stage2;
 
-	/* make sure source component buffer has enough data available and that
-	 * the sink component buffer has enough free bytes for copy. Also
-	 * check for XRUNs.
+	/* Calculate how many blocks can be processed with
+	 * available source and free sink frames amount.
 	 */
-	if (source->avail < need_source) {
-		trace_src_error("src_copy() error: source component buffer "
-				"has not enough data available");
-		return -EIO; /* xrun */
-	}
-	if (sink->free < need_sink) {
-		trace_src_error("src_copy() error: sink component buffer "
-				"has not enough free bytes for copy");
-		return -EIO; /* xrun */
+	if (s2->filter_length > 1) {
```

Comment:
Better to wrap it up into some function.

Reply:
Yep, fixed.